cirq-core/cirq/transformers/routing/mapping_manager.py

```diff
@@ -14,154 +14,196 @@
         """
-        # make sure edge insertion order is the same amongst equivalent graphs.
-        if nx.is_directed(device_graph):
-            self.device_graph = nx.DiGraph()
-            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
-            self.device_graph.add_edges_from(sorted(list(device_graph.edges)))
-        else:
-            self.device_graph = nx.Graph()
-            self.device_graph.add_nodes_from(sorted(list(device_graph.nodes(data=True))))
-            self.device_graph.add_edges_from(
-                sorted(list(sorted(edge) for edge in device_graph.edges))
-            )
-
-        self._map = initial_mapping.copy()
-        self._inverse_map = {v: k for k, v in self._map.items()}
-        self._induced_subgraph = nx.induced_subgraph(self.device_graph, self._map.values())
+        # Map both logical and physical qubits to integers.
+        self._logical_qid_to_int = {q: i for i, q in enumerate(sorted(initial_mapping.keys()))}
+        self._int_to_logical_qid = sorted(
+            self._logical_qid_to_int.keys(), key=lambda x: self._logical_qid_to_int[x]
+        )
+        self._physical_qid_to_int = {q: i for i, q in enumerate(sorted(initial_mapping.values()))}
+        self._int_to_physical_qid = sorted(
+            self._physical_qid_to_int.keys(), key=lambda x: self._physical_qid_to_int[x]
+        )
+        logical_qubits, physical_qubits = (
+            zip(*[(k, v) for k, v in initial_mapping.items()]) if initial_mapping else ([], [])
+        )
+        num_qubits = len(logical_qubits)
+        self._logical_to_physical = np.asarray(
+            [
+                self._physical_qid_to_int[physical_qubits[i]]
+                for i in sorted(
+                    range(num_qubits), key=lambda x: self._logical_qid_to_int[logical_qubits[x]]
+                )
+            ]
+        )
+        self._physical_to_logical = np.asarray(
+            [
+                self._logical_qid_to_int[logical_qubits[i]]
+                for i in sorted(
+                    range(num_qubits), key=lambda x: self._physical_qid_to_int[physical_qubits[x]]
+                )
+            ]
+        )
+        # Construct the induced subgraph (on integers) and corresponding distance matrix.
+        self._induced_subgraph_int = nx.relabel_nodes(
+            nx.induced_subgraph(device_graph, initial_mapping.values()),
+            {q: self._physical_qid_to_int[q] for q in initial_mapping.values()},
+        )
+        # Compute floyd warshall dictionary.
         self._predecessors, self._distances = nx.floyd_warshall_predecessor_and_distance(
-            self._induced_subgraph
+            self._induced_subgraph_int
         )
 
     @property
-    def map(self) -> Dict['cirq.Qid', 'cirq.Qid']:
-        """The mapping of logical qubits (keys) to physical qubits (values)."""
-        return self._map
+    def physical_qid_to_int(self) -> Dict['cirq.Qid', int]:
```

Comment:
`physical_qid_to_int` is never called in `RouteCQC`. Should this still be a property?

Reply:
I think it's useful to still have it there for anyone who wants to debug / inspect the state of their routing given a mapping manager and (potentially custom) transformer. It's also being used in tests.